### DISCUSSION TITLE
Query timing out

### DISCUSSION THREAD

**Discussion by jfevrier10** [moj-analytical-services/splink#1473] [Q&A] [answered]
Currently running:

pairwise_predictions = linker.predict()
clusters = linker.cluster_pairwise_predictions_at_threshold(pairwise_predictions, 0.95)
clusters.as_pandas_dataframe(limit=100)

This section of the tutorial in an Athena based Linker on 17 million records in AWS glue with a large number of columns, but I believe the linker.predict() is taking 30 minutes to run and timing out. I really just wanted to get some results on our dataset to see, I basically copied the Athena Full Example [here](https://github.com/moj-analytical-services/splink/blob/master/tests/test_full_example_athena.py) with data already in AWS Glue

- what improvements could I make to make it not take as long to run?
- can I increase the query timeout value?
- should it take this long when using Athena? 

Thanks!

**Comment by jfevrier10** [comment]
and even the count_num_comparisons_from_blocking_rule times out after 30 minutes:
```
br = "substr(l.first_name,1,3) = substr(r.first_name,1,3) and l.last_name = r.last_name and l.email_address = r.email_address and l.preferred_mailing_city = r.preferred_mailing_city"
linker.count_num_comparisons_from_blocking_rule(br)
```

With it this filtered I figured it would be small enough

**Reply by RobinL** [reply]
That is strange.  You're certainly doing the right thing by trying the `count_num_comparisons_from_blocking_rule ` function.  

We don't use Athena as much as some of the other backends, but I haven't seen this happening before with such a strict blocking rule

As a starting point I would recommend trying DuckDB, and using the new (and much more efficient) function called `_count_num_comparisons_from_blocking_rule_pre_filter_conditions`
https://github.com/moj-analytical-services/splink/blob/22bc640e8fb71ce80e98c04978d971117d201859/splink/linker.py#L2752

So long as that returns a count fewer than (say) 1bn, then Athena should't be struggling to handle it.

At least that might rule out some possibilities for the source of the problem

**Reply by jfevrier10** [reply]
Thanks for the reply, I ran count_num_comparisons again on the rule above and it returned 52,202,023,230 after 25 minutes ! so now I am curious as to how I can filter this out even more to get it under 1 Billion. I assume even switching to Spark wouldn't help with this number of comparisons

**Reply by RobinL** [reply]
Various possibilities come to mind.

- Could it be something to do with nulls not being true nulls in the datasets. If you have loads of rows with blank strings of other values that signify nulls (e.g. the string 'na' or whatever, then that could be the cause.
- Could it be that some entities simply have large numbers of rows associated with them in the original dataset?  e.g. if a single entity had (say) 1,000 records associated with it the input data, then those 1,000 records alone would be responsible for 1 million comparisons.

One way you could look for this is to use the profile columns function to profile each of your blocking columns individually.
```
linker.profile_columns(["substr(first_name, 1,3)", "surname", "email_address", "preferred_mailing_city"])
```

And you can also apply functions like:
```
linker.profile_columns(["concat(first_name, surname, email_address)"])
```

**Reply by jfevrier10** [reply]
Null surname is my top value with 1 million entries. I also have a first name type that I want to exclude from my results for example 'defaultuser' how could I exclude these from my data? I apologize for asking the direct questions I did search the discussion boards

**Reply by RobinL** [reply, marked as answer]
OK - a true null on the surname should be fine.  

One way to check is something like 
```
SELECT
  
  (surname is null) as is_surname_true_null,
*
FROM
  your_table_name
where unique_id = {id of record with null surname here}

```


'defaultuser' could be the problem, especially if the default user has the same email address etc.  You need to do some preprocessing prior to bringing the data into Splink.  

In this case, something like a CASE WHEN statement that has 
```
SELECT *,
       CASE WHEN first_name = 'defaultuser' THEN NULL
            ELSE first_name
       END AS fixed_first_name
FROM df;

```

You can use a CTAS to turn this statement into a materialised table in the Athena catalog, docs [here](https://docs.aws.amazon.com/athena/latest/ug/ctas.html)
